server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java

```diff
@@ -21,214 +21,205 @@
+        if (unconfiguredBootstrapTimeout == null) {
+            return;
+        }
 
         if (transportService.getLocalNode().isMasterNode() == false) {
             return;
         }
 
-        if (unconfiguredBootstrapTimeout != null) {
-            logger.info("no discovery configuration found, will perform best-effort cluster bootstrapping after [{}] " +
-                    "unless existing master is discovered", unconfiguredBootstrapTimeout);
-            final ThreadContext threadContext = transportService.getThreadPool().getThreadContext();
-            try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
-                threadContext.markAsSystemContext();
+        logger.info("no discovery configuration found, will perform best-effort cluster bootstrapping after [{}] " +
+            "unless existing master is discovered", unconfiguredBootstrapTimeout);
+
+        transportService.getThreadPool().scheduleUnlessShuttingDown(unconfiguredBootstrapTimeout, Names.GENERIC, new Runnable() {
+            @Override
+            public void run() {
+                final Set<DiscoveryNode> discoveredNodes = getDiscoveredNodes();
+                final List<DiscoveryNode> zen1Nodes = discoveredNodes.stream().filter(Coordinator::isZen1Node).collect(Collectors.toList());
+                if (zen1Nodes.isEmpty()) {
+                    logger.debug("performing best-effort cluster bootstrapping with {}", discoveredNodes);
+                    startBootstrap(discoveredNodes, emptyList());
+                } else {
+                    logger.info("avoiding best-effort cluster bootstrapping due to discovery of pre-7.0 nodes {}", zen1Nodes);
+                }
+            }
+
+            @Override
+            public String toString() {
+                return "unconfigured-discovery delayed bootstrap";
+            }
+        });
+    }
+
+    private Set<DiscoveryNode> getDiscoveredNodes() {
+        return Stream.concat(Stream.of(transportService.getLocalNode()),
+            StreamSupport.stream(discoveredNodesSupplier.get().spliterator(), false)).collect(Collectors.toSet());
+    }
 
-                transportService.getThreadPool().scheduleUnlessShuttingDown(unconfiguredBootstrapTimeout, Names.SAME, new Runnable() {
+    private void startBootstrap(Set<DiscoveryNode> discoveryNodes, List<String> unsatisfiedRequirements) {
+        assert discoveryNodes.stream().allMatch(DiscoveryNode::isMasterNode) : discoveryNodes;
+        assert discoveryNodes.stream().noneMatch(Coordinator::isZen1Node) : discoveryNodes;
+        assert unsatisfiedRequirements.size() < discoveryNodes.size() : discoveryNodes + " smaller than " + unsatisfiedRequirements;
+        if (bootstrappingPermitted.compareAndSet(true, false)) {
+            doBootstrap(new VotingConfiguration(Stream.concat(discoveryNodes.stream().map(DiscoveryNode::getId),
+                unsatisfiedRequirements.stream().map(s -> BOOTSTRAP_PLACEHOLDER_PREFIX + s))
```

Comment:
add a `-` or something between `BOOTSTRAP_PLACEHOLDER_PREFIX` and requirement? Alternatively, add it directly to the PREFIX.

Reply:
Yep, see b40144f. It already ended in punctuation (i.e. `}`) but a hyphen can't hurt too.